changeset:   23702:d5abf2c2536c
tag:         tip
user:        Blaise Bourdin <bourdin@lsu.edu>
date:        Sun Jun 16 10:16:52 2013 +0200
summary:     Hacked block size into DMMesh

diff -r 93d37d8c07a0 -r d5abf2c2536c include/petscdm.h
--- a/include/petscdm.h	Fri Mar 29 18:06:44 2013 -0500
+++ b/include/petscdm.h	Sun Jun 16 10:16:52 2013 +0200
@@ -121,6 +121,7 @@
 PETSC_EXTERN PetscErrorCode DMGetLocalToGlobalMappingBlock(DM,ISLocalToGlobalMapping*);
 PETSC_EXTERN PetscErrorCode DMCreateFieldIS(DM,PetscInt*,char***,IS**);
 PETSC_EXTERN PetscErrorCode DMGetBlockSize(DM,PetscInt*);
+PETSC_EXTERN PetscErrorCode DMSetBlockSize(DM,PetscInt);
 PETSC_EXTERN PetscErrorCode DMCreateColoring(DM,ISColoringType,const MatType,ISColoring*);
 PETSC_EXTERN PetscErrorCode DMCreateMatrix(DM,const MatType,Mat*);
 PETSC_EXTERN PetscErrorCode DMSetMatrixPreallocateOnly(DM,PetscBool);
diff -r 93d37d8c07a0 -r d5abf2c2536c src/dm/impls/mesh/mesh.c
--- a/src/dm/impls/mesh/mesh.c	Fri Mar 29 18:06:44 2013 -0500
+++ b/src/dm/impls/mesh/mesh.c	Sun Jun 16 10:16:52 2013 +0200
@@ -540,7 +540,7 @@
 PetscErrorCode DMCreateGlobalVector_Mesh(DM dm, Vec *gvec)
 {
   DM_Mesh       *mesh = (DM_Mesh *) dm->data;
-  PetscInt       localSize, globalSize;
+  PetscInt       localSize, globalSize, blocksize;
   PetscErrorCode ierr;
 
   PetscFunctionBegin;
@@ -562,8 +562,10 @@
     localSize  = order->getLocalSize();
     globalSize = order->getGlobalSize();
   }
+  ierr = DMGetBlockSize(dm,&blocksize);CHKERRQ(ierr);
   ierr = VecCreate(((PetscObject) dm)->comm, gvec);CHKERRQ(ierr);
   ierr = VecSetSizes(*gvec, localSize, globalSize);CHKERRQ(ierr);
+  ierr = VecSetBlockSize(*gvec,blocksize);CHKERRQ(ierr);
   ierr = VecSetFromOptions(*gvec);CHKERRQ(ierr);
   ierr = PetscObjectCompose((PetscObject) *gvec, "DM", (PetscObject) dm);CHKERRQ(ierr);
   PetscFunctionReturn(0);
@@ -574,7 +576,7 @@
 PetscErrorCode DMCreateLocalVector_Mesh(DM dm, Vec *lvec)
 {
   DM_Mesh       *mesh = (DM_Mesh *) dm->data;
-  PetscInt       size;
+  PetscInt       size,blocksize;
   PetscErrorCode ierr;
 
   PetscFunctionBegin;
@@ -592,8 +594,10 @@
     if (!flag) SETERRQ(((PetscObject) dm)->comm,PETSC_ERR_ARG_WRONGSTATE, "Must set default section");
     size = m->getRealSection("default")->getStorageSize();
   }
+  ierr = DMGetBlockSize(dm,&blocksize);CHKERRQ(ierr);
   ierr = VecCreate(PETSC_COMM_SELF, lvec);CHKERRQ(ierr);
   ierr = VecSetSizes(*lvec, size, size);CHKERRQ(ierr);
+  ierr = VecSetBlockSize(*lvec,blocksize);CHKERRQ(ierr);
   ierr = VecSetFromOptions(*lvec);CHKERRQ(ierr);
   ierr = PetscObjectCompose((PetscObject) *lvec, "DM", (PetscObject) dm);CHKERRQ(ierr);
   PetscFunctionReturn(0);
diff -r 93d37d8c07a0 -r d5abf2c2536c src/dm/interface/dm.c
--- a/src/dm/interface/dm.c	Fri Mar 29 18:06:44 2013 -0500
+++ b/src/dm/interface/dm.c	Sun Jun 16 10:16:52 2013 +0200
@@ -609,6 +609,29 @@
 }
 
 #undef __FUNCT__  
+#define __FUNCT__ "DMSetBlockSize"
+/*@
+   DMSetBlockSize - Sets the inherent block size associated with a DM
+
+   Not Collective
+
+   Input Parameter:
+.  dm - the DM with block structure
+.  bs - the block size, 1 implies no exploitable block structure
+
+   Level: intermediate
+
+.seealso: ISCreateBlock(), VecSetBlockSize(), MatSetBlockSize(), DMGetLocalToGlobalMappingBlock()
+@*/
+PetscErrorCode  DMSetBlockSize(DM dm,PetscInt bs)
+{
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(dm,DM_CLASSID,1);
+  dm->bs = bs;
+  PetscFunctionReturn(0);
+}
+
+#undef __FUNCT__  
 #define __FUNCT__ "DMCreateInterpolation"
 /*@
     DMCreateInterpolation - Gets interpolation matrix between two DMDA or DMComposite objects

